Clean up console access code

smurfix opened the pull request:
#### Brief overview of PR changes/additions

Nothing brief about this I'm afraid.

* added `WINDOW_NAME` and `CONSOLE` macro calls to `TLuaInterpreter`; rationale: make code more readable, unify error messages
* moved a heap of `mudlet::X(Host &, …)` calls to `Host::X(…)`; rationale: drop a level of indirection, code readability
* removed `QMap<Host*, X>` variables from `mudlet::`, replaced with data fields in `Host`; rationale: anti-pattern, may leak references, drop a level of indirection, readbilty
* remove `mudlet::mTabMap`; redundant
* add an iterator to `HostMap`; previously a list of names was generated only to look up each name in `mHostMap`

#### Motivation for adding to Mudlet

Removing these historically-grown code artefacts is fun, uncovers a latent bug or two, and paves the way forward.

#### Other info (issues closed, discussion etc)

I believe that these patches do not introduce any change in behavior. That will happen in the next part (filter name collisions).

smurfix committed:
Add a new Host::findConsole function

smurfix committed:
Sketch: TLuaInterpreter.cpp modification

smurfix committed:
Fix macros to eval their argument once

smurfix committed:
More adaptions

smurfix committed:
TLuaInterpreter: use WINDOW_NAME macro.

Some cases not touched because difficult.

smurfix committed:
macro-ify a lot of functions

smurfix committed:
Remove no-longer-needed accessors

smurfix committed:
Clean up font and console handling

* unify setFontSize and setMiniConsoleFontSize
  the latter is now an alias to the former
* TConsole methods don't need "MiniConsole" in their name
  that's redundant
* teach Host::updateConsolesFont() to refresh the main console
  no need to expose these details

smurfix committed:
Fix font setting

smurfix committed:
Removed a bunch more unused methods from mudlet.cpp

smurfix committed:
replace HostManager::getHostList

It was only used to
* create a list to test membership against
  just call getHost and check for null
* create a list of host names, each of which is then looked up
  which is unnecessary work

smurfix committed:
Remove the "QMap<Host*, T*>" anti-pattern

It makes much more sense to add a T* member to the Host class.

smurfix committed:
Allow window names beyond the Lua stack

simply treat a missing parameter (or nil) as empty string

smurfix committed:
Use a nested class for iteration

smurfix committed:
Fix host iterator

smurfix committed:
move getLines() to Host

smurfix committed:
remove mudlet::mTabMap

Redundant, the host map and Host::mpConsole works.

smurfix committed:
more moves to host

smurfix committed:
Move host-close code to the host

smurfix committed:
Move even more stuff from mudlet:: to THost::

Also drop mudlet::mpProfilePreferencesDlgMap in favor of a pointer in
THost. Sorry that this got mashed in but I was halfway through this
before I noticed and didn't want to do the work twice.

smurfix committed:
These things are pointers, thus should be named as such

as annoying as that is …

smurfix committed:
Vanquish another two pointer-to-host maps

smurfix committed:
Move host-specific layout updates to THost

This patch also removes layering violations in TToolbar and TDockWidget.

smurfix committed:
mudlet.h include directives cleanup

smurfix committed:
appease CodeFactor

smurfix committed:
variable rename

smurfix committed:
Check for valid dialog

smurfix committed:
minor cleanup

smurfix committed:
Revert mistaken checkin of test code

from d918164c77382332ad1cd9bd31b2d1ea173267fd

smurfix committed:
rename function in error messages

smurfix committed:
Crash fix

smurfix committed:
renamed args for clarity

smurfix committed:
Braces please

smurfix committed:
Bugfix: "echo" didn't talk to labels

smurfix committed:
Found two wrongly-named exceptions

smurfix committed:
Clean up merge differences

SlySven commented:
I'm not sure I can see any difference from #4167 and **that** got *reverted* #4179 because it broke some things.

I'm sorry (and I am sometimes guilty of the exactly the same crime) but I really think the others want smaller bite size PRs that are easier to analysis and digest. 🙁 

Are there any things that can be broken out and done on their own? 🥺

smurfix commented:
The fix for the expiration counter problem has been merged independently.
The fix for "lua echo("someLabelName", "content") is included.
Both really are in there, I just checked.
As I don't know of any other bug there are no other differences. Should there be?

Yes sorry this is big; I promise to do better in the future.
For easier review you might want to look at the individual diffs instead.

vadi2 commented:
I remember, the only other known issue are the Coverity warnings and the crash opening settings w/o a profile.

Will continue testing to find anything else.

smurfix committed:
Merge remote-tracking branch 'origin/development' into subconsole3

smurfix committed:
Merge error

vadi2 commented:
This needs to be updated again

smurfix committed:
Merge remote-tracking branch 'origin/development' into subconsole3

smurfix commented:
Updated.

I also requested Github to implement a notification for "new" merge conflicts. We'll see what they say about that.

vadi2 commented:
Are the Coverity complaints from the initial merge attempt still outstanding in this PR?

smurfix commented:
I will check.

smurfix committed:
Fix potential NULL deref

smurfix commented:
One has been fixed already, one I just fixed, the others look like they are false positives.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -2391,6 +2391,9 @@ void mudlet::show_action_dialog()
 void mudlet::show_options_dialog(const QString& tab)
 {
     Host* pHost = getActiveHost();
+    if (!pHost) {
```
I think this might be a regression, because the settings window is openable w/o any host - for the edge case that you'd want to change the language Mudlet is in before connecting...

smurfix replied:
I'll check.

smurfix replied:
Ah, yes, the old code seems to add a NULL pointer to the `mpProfilePreferencesDlgMap` … which was never cleared anyway, so interesting dangerous special effects either way.

I'll fix that.

smurfix replied:
Cleanup pushed. The pointer to the dialog is now shared not scoped, otherwise I'd have to duplicate the code in question.

smurfix committed:
Re-add the option to have a prefs dialog without a host

vadi2 commented:
Unfortunately it still makes a :boom:, do this:

1) open Mudlet, don't open a profile
2) open settings, change the language to German, hit save
3) open settings once again

```
1   QTabBar::setCurrentIndex(int)                                        0x7ffff78491c2 
2   dlgProfilePreferences::setTab         dlgProfilePreferences.cpp 1449 0x5555557fda0f 
3   mudlet::show_options_dialog           mudlet.cpp                2416 0x55555596a7bb 
4   mudlet::slot_show_options_dialog      mudlet.cpp                2505 0x55555596b94e 
```

This seems to be specific to this branch.

smurfix commented:
Bah. The dialog now gets destroyed even though the code clearly expects, and always expected, that not to happen.

Investigating.

smurfix committed:
Don't delete the prefs dialog on close

smurfix commented:
Don't you just *love* hidden bugs to surface when you're busy fixing things that don't seem to have anything to do with them?

vadi2 commented:
That's the price to pay 😉

Will continue reviewing this later when time permits. Anyone else is also welcome to help in testing!

smurfix committed:
Fix "Don't delete the prefs dialog on close" another way.

Apparently Qt correctly handles this when using a QScopedPointer but not
using a QSharedPointer. That is, even though the shared pointer still
references the dialog, it's killed off without clearing the pointer.

Since keeping the dialog around forever isn't a good idea, replace with
a QPointer which *is* cleared.

smurfix commented:
Apparently Qt correctly handles this when using a QScopedPointer but not using a QSharedPointer. That is, using a QSharedPointer, when the dialog is deleted the pointer to it isn't.

Since keeping the dialog around forever doesn't make much sense the current patch fixes this by using a simple QPointer, which *is* cleared.

vadi2 commented:
Things are a bit busy with the release, will look at testing this more when time permits. Could you update it?

smurfix commented:
In progress.

smurfix committed:
Merge remote-tracking branch 'origin/development' into subconsole3

smurfix commented:
Pushed.

vadi2 commented:
@demonnic could you use the Lua tests on this?

demonnic commented:
good news on that front

![image](https://user-images.githubusercontent.com/3660/98010080-49918080-1dc4-11eb-8f9d-9e278390e280.png)

vadi2 committed:
Merge remote-tracking branch 'origin/development' into subconsole3
